### ISSUE TITLE
Feature Gate: Validator applies cost tracker to blocks during replay

### ISSUE THREAD

**Issue by taozhu-chicago** [solana-labs/solana#29595]
### Description

Blocks exceed cost tracker limits (eg. block limit, account limit etc) should fail during replay with error. 

(This issue is to track an already implemented feature gate, that requires refreshment before activation)

### Feature ID

2ry7ygxiYURULZCrypHhveanvP5tzZ4toRwVp89oCNSj

### Activation Method

Single Core Contributor

**Comment by taozhu-chicago** [comment]
v1.15+ have bunch of changes to cost model, including use requested CU (#27608), add cost to more built-in (#29795), to name a few. This feature should only be activated when cluster is at 1.15+

**Comment by taozhu-chicago** [comment]
Also IMPORTANT that to enable this feature _AFTER_ [QoS adjustment logic](https://github.com/solana-labs/solana/blob/master/core/src/banking_stage/consumer.rs#L427-L432) is removed.

**Comment by apfitzge** [comment]
> Also IMPORTANT that to enable this feature AFTER [QoS adjustment logic](https://github.com/solana-labs/solana/blob/master/core/src/banking_stage/consumer.rs#L427-L432) is removed, https://github.com/solana-labs/solana/issues/31379

I think before either of these are activated we need to have some punishing or incentive to request accurate CUs. Otherwise we will kill block size immediately I think?

**Comment by taozhu-chicago** [comment]
> > Also IMPORTANT that to enable this feature AFTER [QoS adjustment logic](https://github.com/solana-labs/solana/blob/master/core/src/banking_stage/consumer.rs#L427-L432) is removed, #31379
> 
> I think before either of these are activated we need to have some punishing or incentive to request accurate CUs. Otherwise we will kill block size immediately I think?

Yes, #31453 is first step toward to it, hope those actively utilizing priority will first revise their requested CU to be more accurate. If result is positive, can try next steps with SIMD #17

**Comment by apfitzge** [comment]
https://dune.com/crypto_notte/solana-additional-fee-analysis shows we have roughly 30-60% priority fee usage. I think costs also check for builtins to filter those out, unsure what percentage of txs are just builtins. We probably still have some percentage of non-builtin txs that aren't using priority fees, and would use the default 200k cost. Would be great to get a relative idea on how many non-builtin, non-prioritized txs we actually see, since those are the troublemakers w/ removing the cost adjustment post-execution.

hehe that link for SIMD doesn't work ([correct link](https://github.com/solana-foundation/solana-improvement-documents/pull/19)) - also it's really SIMD 19 😆

**Comment by taozhu-chicago** [comment]
> https://dune.com/crypto_notte/solana-additional-fee-analysis shows we have roughly 30-60% priority fee usage. I think costs also check for builtins to filter those out, unsure what percentage of txs are just builtins. We probably still have some percentage of non-builtin txs that aren't using priority fees, and would use the default 200k cost. Would be great to get a relative idea on how many non-builtin, non-prioritized txs we actually see, since those are the troublemakers w/ removing the cost adjustment post-execution.
> 

#31755 documents fee details by using ledger-tool. Welcome to comment there.

**Comment by buffalu** [comment]
Blocked on https://github.com/solana-labs/solana/pull/32285
